### PR TITLE
Upstream changes from cheese3d including GPU acceleration

### PR DESCRIPTION
**Pull request by darsnack**
Hi @lambdaloop apologies for the delay, but as previously discussed, this PR should upstream changes made to aniposelib as part of developing cheese3d. The two major sources of changes are:
- adding Jax as a dependency and updating initial triangulation to use it for acceleration (GPU, falls back to CPU when unavailable)
- updating calibration to support OpenCV 4.7+ APIs (the current main branch uses the latest package version but potentially deprecated APIs)

~~Possibly this needs a rebase before it can be merged, but I am opening the PR to help guide that process.~~ Rebasing complete.

### CODE REVIEW

**Comment by darsnack** [comment]
Okay I think this is ready to review!

**Comment by darsnack** [comment]
cc @HelenHouLab @helenxhou

**Comment by lambdaloop** [comment]
Thank you! I looked through the code and it looks good. I want to run some tests on videos before I merge, I hope to get this before the end of the week.

**Comment by lambdaloop** [comment]
Hello, I ran a test on the triangulation, and it has more error compared to the reference implementation. 

Here are some example files:
https://drive.google.com/drive/folders/1IpHFeFtPfjoWHw1mO-3-qfvz_G5uEovx?usp=sharing

When I run this script (included as `test_ref.py` in the folder above), I see the following output:
```python
from aniposelib.cameras import CameraGroup
import numpy as np

cgroup = CameraGroup.load('calib_motobs_7cam.toml')
d = np.load('point_cloud.npz')
p3d_ref = d['p3d'] # true 3d point
p2d = d['p2d'] # projected 2d points

d2 = np.load('results_classic.npz')
p3d_classic = d2['p3d_tri'] # 3d points from reference aniposelib

# triangulate with jax
p3d_new = cgroup.triangulate(p2d)
reproj = cgroup.reprojection_error(p3d_new, p2d, mean=True)
p3d_new[reproj > 3] = np.nan # triangulate failed (this is expected for some points here)

# reference error 
error_ref = np.linalg.norm(p3d_classic - p3d_ref, axis=1)
print("max reference error:", np.nanmax(error_ref))

error = np.linalg.norm(p3d_new - p3d_ref, axis=1)
print("max error of jax:", np.nanmax(error))
```


Output:
```
max reference error: 0.0011253452829623947
max error of jax: 0.1422341302498858
```

It seems the jax implementation has a higher error for triangulating points than the reference implementation. 
Would you be able to debug this?

**Comment by lambdaloop** [comment]
Some guesses on what it could be:
* possibly that float64 precision is not being used?
* something to do with the jax svd implementation
* something to do with the way you take care of nans

**Comment by lambdaloop** [comment]
For the other changes, I ran a test on calibration using the updated boards.py and it works well, thank you!! 

I'll have to wait on fixes to the accuracy of the jax triangulation before merging this though.

(Note also I'll be out on vacation for the next 2 weeks. Will check back in on March 23rd.)

**Comment by darsnack** [comment]
Thanks, I'll take a look at the error issues and report back.

**Comment by darsnack** [comment]
I looked into this, and I was able to reproduce the error you saw. 

For the fix, your first guess was correct. Enabling 64-bit precision makes the two cases match.
```
➜ JAX_ENABLE_X64=1 python test_ref.py
max reference error: 0.0011253452829623947
max error of jax: 0.0011253452829062308
```

I would say since 32-bit is standard for any GPU compute, this is a trade-off paid for acceleration. We could add a note somewhere about this and the fix, but I think most people will not mind the issue since Anipose will optimize the initial triangulation anyways.

**Comment by lambdaloop** [comment]
Hello I'm back! Hmm, I don't like a change messing with the accuracy of the triangulation still... 

Would it be possible to default to 64 bit precision enabled? I see that we can run something like:
```python
jax.config.update("jax_enable_x64", True)
``` 

When developing anipose, I had some issues (in some datasets) with the bundle adjustment converging when it was run in 32 bit precision, whereas it worked well in 64 bit precision. So I'm nervous to have 32 bit precision as the default... There are likely sensitive cameras out there that would benefit from the 64 bit precision.

If you like, you could add an argument to cgroup.triangulate for 32bit or lower precision for potentially faster inference at the cost of accuracy.

**Comment by darsnack** [comment]
Sure, let's figure out a good way to support toggling 64-bit.

The issue is that almost all GPUs have hardware designed for 32-bits natively. So, at best, 64-bit will result in twice the memory usage and potentially 2x or more latency for any operation. At worst, the hardware simply won't support 64-bits and either crash or silently truncate.

Separately from this, jax itself allows enabling 64-bit either through `jax.config.enable_x64` or the environment variable I used in my test. But the way [this mechanism is designed](https://github.com/jax-ml/jax/issues/22688#issuecomment-2253298306) means that it will be enabled permanently for the rest of the process. That would present an issue if the program uses jax elsewhere outside of Anipose (e.g. as we might in Cheese3D).

I see a couple options, let me know what you think is best:
1. `jax` is not installed by default and numpy is used everywhere (so 64-bit). If the user opts-in by installing `jax` themselves (or we have an extra like `pip install anipose[jax]`), we use `jax` for initial triangulation with 32-bit. At the loss of precision and subject to the caveats you notice, the user can get some speed-up.
2. We install and use jax, but have a flag to toggle 64-bit precision that is enabled by default. Here we can either:
    
    a. Use an environment variable like `ANIPOSE_JAX_DISABLE_64` which if set true, will skip `jax.config.enable_x64`. This would only require changes to aniposelib.
    
    b. Use a keyword passed to `cgroup.triangulate` but that would require changes to anipose as well for threading that option all the way through.

I'm open to implementing any of these options. I think (1) is preferable since a user will by default get the higher precision pathway. And to know they can explicitly opt-in, they will need to read the docs and see the note about precision. Potentially I can also add a warning hinting about the issue to convergence failures as well. Additionally, once a user does opt-in, they get the safest accelerated pathway on a GPU.

**Comment by lambdaloop** [comment]
Hmm that's true it will use 2x the memory and potentially 2x the time. I'd argue that memory isn't as much of a concern for triangulating points, but it is annoying to have 2x the time.. Still, I don't want a change that might mess with the accuracy by default. 

Option 1 could work but I'm worried about triangulation breaking if someone installs jax for an unrelated reason. We would also need to add if statements for numpy and jax.

I would prefer option 2 if that's alright. I think we could implement both options that you specified, it's reasonably easy. For users of aniposelib, I feel that a parameter to cgroup.triangulate is the cleanest. Then you could use the ANIPOSE_JAX_DISABLE_64 argument for your current code, and we could add arguments to support 32 bit operations in anipose in the future.

**Comment by darsnack** [comment]
Okay, I've implemented the flags. Here is an output from testing them:
```
max reference error: 0.0011253452829623947
max error of jax (64-bit): 0.0011253452829062308
max error of jax (32-bit flag): 0.1094434498084814
max error of jax (32-bit ENV VAR): 0.1094434498084814
```
with the updated test script:
```python
#!/usr/bin/env python3
from aniposelib.cameras import CameraGroup
import numpy as np
import multiprocessing as mp

cgroup = CameraGroup.load('calib_motobs_7cam.toml')
d = np.load('point_cloud.npz')
p3d_ref = d['p3d'] # true 3d point
p2d = d['p2d'] # projected 2d points

d2 = np.load('results_classic.npz')
p3d_classic = d2['p3d_tri'] # 3d points from reference aniposelib

def triangulate_worker_64(p2d_data):
    """Worker for 64-bit triangulation"""
    from aniposelib.cameras import CameraGroup
    cgroup = CameraGroup.load('calib_motobs_7cam.toml')
    return cgroup.triangulate(p2d_data)

def triangulate_worker_32_flag(p2d_data):
    """Worker for 32-bit triangulation via flag"""
    from aniposelib.cameras import CameraGroup
    cgroup = CameraGroup.load('calib_motobs_7cam.toml')
    return cgroup.triangulate(p2d_data, disable_64bit=True)

def triangulate_worker_32_env(p2d_data):
    """Worker for 32-bit triangulation via environment variable"""
    import os
    os.environ["ANIPOSE_DISABLE_JAX_X64"] = "1"
    from aniposelib.cameras import CameraGroup
    cgroup = CameraGroup.load('calib_motobs_7cam.toml')
    return cgroup.triangulate(p2d_data)


def _run_in_fresh_process(worker, p2d_data):
    """Run worker in a fresh process and return result."""
    ctx = mp.get_context("spawn")
    with ctx.Pool(processes=1, maxtasksperchild=1) as pool:
        return pool.apply(worker, (p2d_data,))

# Run triangulations in separate processes
if __name__ == '__main__':
    p3d_new_32 = _run_in_fresh_process(triangulate_worker_32_flag, p2d)
    p3d_new_32_env = _run_in_fresh_process(triangulate_worker_32_env, p2d)
    p3d_new_64 = _run_in_fresh_process(triangulate_worker_64, p2d)

    # Calculate reprojection errors
    reproj = cgroup.reprojection_error(p3d_new_64, p2d, mean=True)
    reproj_32 = cgroup.reprojection_error(p3d_new_32, p2d, mean=True)
    reproj_32_env = cgroup.reprojection_error(p3d_new_32_env, p2d, mean=True)

    p3d_new_64[reproj > 3] = np.nan # triangulate failed (this is expected for some points here)
    p3d_new_32[reproj_32 > 3] = np.nan
    p3d_new_32_env[reproj_32_env > 3] = np.nan

    # reference error
    error_ref = np.linalg.norm(p3d_classic - p3d_ref, axis=1)
    print("max reference error:", np.nanmax(error_ref))

    error = np.linalg.norm(p3d_new_64 - p3d_ref, axis=1)
    print("max error of jax (64-bit):", np.nanmax(error))
    error = np.linalg.norm(p3d_new_32 - p3d_ref, axis=1)
    print("max error of jax (32-bit flag):", np.nanmax(error))
    error = np.linalg.norm(p3d_new_32_env - p3d_ref, axis=1)
    print("max error of jax (32-bit ENV VAR):", np.nanmax(error))
```

**Comment by lambdaloop** [comment]
Alright that works, thank you! I'll merge this now!!

**Comment by darsnack** [comment]
Would it be possible to make a new release of anipose and aniposelib with these changes in place? Having a pypi release that we can point to is the only thing preventing us from releasing cheese3d.

Thanks!

**Comment by lambdaloop** [comment]
Ah yes, I'll do it now!

**Comment by darsnack** [comment]
Awesome, thanks for such a quick response!

**Comment by lambdaloop** [comment]
Okay it's pushed as aniposelib 0.8.0
